src/Octoshift/Services/GithubApi.cs

```diff
@@ -1089,24 +1090,102 @@ public virtual async Task<string> UploadArchiveToGithubStorage(string orgDatabas
+            return (string)data["uri"];
         }
+    }
 
-        var data = JObject.Parse(response);
-        return (string)data["uri"];
+    private async Task<string> UploadMultipart(Stream archiveContent, string archiveName, string uploadUrl)
+    {
+        var buffer = new byte[_streamSizeLimit];
+
+        // 1. Start the upload
+        var (startResponse, startHeaders) = await StartUploadAsync(uploadUrl, archiveName, archiveContent.Length);
+
+        var nextUrl = GetNextUrl(startHeaders);
+        var guid = System.Web.HttpUtility.ParseQueryString(nextUrl.Query)["guid"];
+
+        // 2. Upload parts
+        int bytesRead;
+        while ((bytesRead = await archiveContent.ReadAsync(buffer, 0, buffer.Length)) > 0)
+        {
+            nextUrl = await UploadPartAsync(buffer, bytesRead, nextUrl.ToString());
+        }
+
+        // 3. Complete the upload
+        await CompleteUploadAsync(nextUrl.ToString());
+
+        return $"gei://archive/{guid}";
+    }
+
+    private async Task<(string Response, KeyValuePair<string, IEnumerable<string>>[] Headers)> StartUploadAsync(string uploadUrl, string archiveName, long contentSize)
+    {
+        // Create the body for the request
+        var body = new JObject
+        {
+            ["content_type"] = "application/octet-stream",
+            ["name"] = archiveName,
+            ["size"] = contentSize
+        };
+
+        // Post the request and get both response content and headers as a tuple
+        var (response, headers) = await _client.PostTupleAsync(uploadUrl, body);
+
+        // Return the tuple with response and headers
+        return (response, headers);
+    }
+
+
+    private async Task<Uri> UploadPartAsync(byte[] body, int bytesRead, string nextUrl)
+    {
+        var content = new ByteArrayContent(body, 0, bytesRead);
+        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/octet-stream");
+        var (response, headers) = await _client.PatchTupleAsync(nextUrl, content);
+
+        return GetNextUrl(headers);
+    }
+
+    private async Task CompleteUploadAsync(string lastUrl)
+    {
+        // Create an empty HttpContent object
+        var content = new StringContent(string.Empty);
+
+        // Set the correct Content-Type header (adjust based on API requirements)
+        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/octet-stream");
+
+        await _client.PutAsync(lastUrl, content);
+    }
+
+    private Uri GetNextUrl(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+    {
+        // Find the Location header
+        var locationHeader = headers.FirstOrDefault(header => header.Key == "Location");
+
+        // Check if the Location header exists and has a value
+        if (locationHeader.Key != null)
+        {
+            var locationValue = locationHeader.Value.FirstOrDefault();
+
+            if (!string.IsNullOrEmpty(locationValue))
```

Comment:
We do have a handy extension method for this:
```suggestion
            if (locationValue.HasValue())
```

Reply:
![Screenshot 2024-10-23 at 12 53 37 PM](https://github.com/user-attachments/assets/2457d8ec-382f-41d4-9293-dc2dfc9b631f)

Reply:
Its not letting me do that

Reply:
that's because you need to add the appropriate `using` to the beginning of the cs file (`using OctoshiftCLI.Extensions`). It's an extension method that originally doesn't belong to the `string` class.